### PR TITLE
Integrate quota with Alluxio Cache

### PR DESCRIPTION
**Pull request by apc999**
Test plan - Unit test added

== RELEASE NOTES ==

Hive Changes
* Support Cache Quota when using Alluxio cache

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1097,6 +1097,14 @@
                 <artifactId>alluxio-shaded-client</artifactId>
                 <version>${dep.alluxio.version}</version>
                 <exclusions>
+                    <exclusion>
```
Comment (apc999) [inline pom.xml:1100]: please ignore change in this file, just to make the build work with Alluxio client on snapshot versions.
once release version is cut, will revert the change

File: presto-cache/pom.xml [hunk 2]
```diff
@@ -133,4 +133,24 @@
             <scope>test</scope>
         </dependency>
     </dependencies>
+
```
Comment (apc999) [inline presto-cache/pom.xml:136]: please ignore change in this file, just to make the build work with Alluxio client on snapshot versions.
once release version is cut, will revert the change

File: presto-cache/pom.xml [hunk 3]
```diff
@@ -133,4 +133,24 @@
...
+    <build>
```
Comment (jainxrohit) [inline presto-cache/pom.xml:137]: This seems unrelated? Do we need this?

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java [hunk 4]
```diff
@@ -79,6 +85,16 @@ public FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
...
                     .setFolder(false)
                     .setLength(hiveFileContext.getFileSize().get());
+            if (cacheQuotaEnabled) {
+                CacheScope scope = CacheScope.create(hiveFileContext.getCacheQuota().getIdentity());
+                info.setCacheScope(scope);
+                if (hiveFileContext.getCacheQuota().getQuota().isPresent()) {
+                    info.setCacheQuota(new CacheQuota(ImmutableMap.of(scope.level(), hiveFileContext.getCacheQuota().getQuota().get().toBytes())));
+                }
+                else {
+                    info.setCacheQuota(CacheQuota.UNLIMITED);
```
Comment (shixuan-fan) [inline presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:95]: IIUC unlimited quota is the default behavior. How about:
```
if (cacheQuotaEnabled && hiveFileContext.getCacheQuota().getQuota().isPresent()) {
  // do cache quota magic
}
```

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java [hunk 5]
```diff
@@ -63,6 +68,7 @@ public synchronized void initialize(URI uri, Configuration configuration)
                 throw new IOException("Failed to open file", e);
             }
         });
+        this.cacheQuotaEnabled = configuration.getBoolean(PropertyKey.USER_CLIENT_CACHE_QUOTA_ENABLED.getName(), false);
```
Comment (shixuan-fan) [inline presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:71]: nit: can we static import `USER_CLIENT_CACHE_QUOTA_ENABLED`?

**Review by shixuan-fan — APPROVED** [review]
LGTM % minor comments. It would be great if @kewang1024 could also take a pass :)

File: presto-cache/pom.xml (hunk 3 above)
Comment (apc999) [reply presto-cache/pom.xml:137]: explained above. for some reason on snapshot alluxio, enforcer will fail on build

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java (hunk 5 above)
Comment (apc999) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:71]: done

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java (hunk 4 above)
Comment (apc999) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:95]: not default in Alluxio. so I set it explicitly.
note that this is the data structure passed to Alluxio land

File: pom.xml (hunk 1 above)
Comment (jainxrohit) [reply pom.xml]: These changes are not sufficient, it seems more work is needed to manage proper dependencies. The newer version is bringing incompatible version of many dependencies.

Comment (jainxrohit) [reply pom.xml]: The build in the PR as well is failing.

**Comment by apc999** [comment]
@kewang1024 @jainxrohit  PTAL

**Comment by apc999** [comment]
@kewang1024 @jainxrohit ping

**Comment by apc999** [comment]
Pingg @kewang1024

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java (hunk 4 above)
Comment (kewang1024) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:95]: Question: what is the default behavior if not set?

**Review by kewang1024 — APPROVED** [review]

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java (hunk 4 above)
Comment (apc999) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:95]: Quota check will never fail. The behavior will be exactly the same as today

Comment (kewang1024) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java:95]: then shouldn't it be the same as UNLIMITED?

**Comment by apc999** [comment]
@kewang1024 the failing test is irrelevant. Please advice 

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.0.1:jar (attach-javadocs) on project presto-kudu: MavenReportException: Error while generating Javadoc: artifact resolver problem - Failed to read artifact descriptor for org.apache.yetus:audience-annotations:jar:0.8.0: Could not transfer artifact org.apache.yetus:audience-annotations:pom:0.8.0 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/apache/yetus/audience-annotations/0.8.0/audience-annotations-0.8.0.pom: Connection timed out (Read failed) -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :presto-kudu
Error: Process completed with exit code 1.
```
